3D Trade.

ChrisEddy opened the pull request.

avaer commented on components/App.js:
```diff
@@ -7,7 +7,7 @@ import {planet} from '../planet.js';
 import {setBindings} from './bindings.js';
 import {getContractSource} from '../blockchain.js';
-import DiffDOM from '../diffDOM.js';
+import DiffDOM from '../diffDOM.js'; 
```
Extra space.

avaer commented on components/App.js:
```diff
@@ -42,6 +42,124 @@ export const onclickBindings = {
     });
   },
+  'threeD-social-tradeBtn': (e) => {
+    const { menu } = getState();
```
Bad brackets.

ChrisEddy replied:
what bad brackets? looks good to me.

avaer reviewed (changes requested):
IDE is not following linting rules :(

avaer commented on components/App.js:
```diff
@@ -42,6 +42,124 @@ export const onclickBindings = {
+  },
+  'threeD-trade-inventory-forward': (e) => {
+    const { menu } = getState();
+    ++menu.trade.inventoryPage;
```
Our convention is postfix `++`.

avaer replied:
Also, is this breaking past the total number of pages of content? It could be capped.

ChrisEddy replied:
fixed

avaer commented on components/App.js:
```diff
@@ -42,6 +42,124 @@ export const onclickBindings = {
+  },
+  'threeD-trade-peers-back': (e) => {
+    const { menu } = getState();
+    menu.trade.peersPage > 0 ? --menu.trade.peersPage : null;
```
This should probably be `Math.max`

ChrisEddy replied:
no

avaer commented on components/App.js:
```diff
@@ -42,6 +42,124 @@ export const onclickBindings = {
+  },
+  'threeD-trade-inventory-back': (e) => {
+    const { menu } = getState();
+    menu.trade.inventoryPage > 0 ? --menu.trade.inventoryPage : null;
```
`Math.max`?

avaer commented on components/App.js:
```diff
@@ -42,6 +42,124 @@ export const onclickBindings = {
+      const response2 = await res.json();
+      console.log(response2)
+
+      menu.trade = {
```
This looks like it should be a piecewise assignment, not a replacement.

Otherwise this code and the initialization code could become out of sync.

avaer commented on threeD-components/Inventory.js:
```diff
@@ -183,7 +184,8 @@ export const Inventory = (props = {}) => {
         border: 5px solid;
-        padding: 10px;
+        padding: 20px;
+        margin: 20px 0px 20px 0px;
```
Prefer this to be explicit direction overrides, not a circle ordering.

ChrisEddy replied:
git gud

avaer replied:
I don't understand.

avaer commented on threeD-components/Social.js:
```diff
@@ -56,7 +56,8 @@ const Social = (props = {}) => {
       <div class=peer-list>
         ${peers.length > 0 ? peers.map(rig => {
-          return `<a class=peer id=peer name=${rig.peerConnection.connectionId}>
+          console.log(rig)
```
Extra logging.

avaer commented on threeD-components/Social.js:
```diff
@@ -56,7 +56,8 @@ const Social = (props = {}) => {
         ${peers.length > 0 ? peers.map(rig => {
-          return `<a class=peer id=peer name=${rig.peerConnection.connectionId}>
+          console.log(rig)
+          return `<a class=peer id=peer name=${rig?.peerConnection?.connectionId}>
```
When will rig be falsey?

avaer commented on vr-ui.js:
```diff
@@ -1485,6 +1485,7 @@ const makeMenuMesh = (cubeMesh, onclickBindings) => {
       allItems: state.menu.browse.items,
       peers: state.menu.world.peers,
       selectedPeerId: state.menu.world.selectedPeerId,
+      trade: state.menu.trade
```
Needs a trailing comma.

avaer commented on components/App.js:
```diff
@@ -102,8 +220,10 @@ export const onclickBindings = {
       visible: false,
```
We should not reset such a large object; simply clear the values we need to clear. Otherwise this has a dependency on the structure initializer, which is hidden to the programmer.

avaer commented on components/App.js:
```diff
@@ -42,6 +42,124 @@ export const onclickBindings = {
+  },
+  'threeD-trade-cancel': (e) => {
+    const { menu } = getState();
+    menu.trade = {
```
Another structure initialization dependency.

avaer commented on components/App.js:
```diff
@@ -42,6 +42,124 @@ export const onclickBindings = {
+  },
+  'threeD-trade-header-backBtn': (e) => {
+    const { menu } = getState();
+    menu.trade = {
```
Another structure initialization dependency.

avaer commented on components/App.js:
```diff
@@ -42,6 +42,124 @@ export const onclickBindings = {
+        peersPage: 0,
+      }
+    } else {
+      // no agreement
```
Don't need to set state then right?

ChrisEddy replied:
fixed

avaer commented on threeD-components/Trade.js:
```diff
@@ -0,0 +1,303 @@
+}
+
+const Trade = (props) => {
+  let inventoryItems = props.inventoryItems.slice(3 * props.trade.inventoryPage, 3 * (props.trade.inventoryPage + 1));
```
Why not `const`?

avaer commented on threeD-components/Trade.js:
```diff
@@ -0,0 +1,303 @@
+
+const Trade = (props) => {
+  let inventoryItems = props.inventoryItems.slice(3 * props.trade.inventoryPage, 3 * (props.trade.inventoryPage + 1));
+  let peers = props.peers.slice(3 * props.trade.peersPage, 3 * (props.trade.peersPage + 1));
```
`3` sounds like it should be a constant.

avaer commented on threeD-components/Trade.js:
```diff
@@ -0,0 +1,303 @@
+            </h1>
+          </div>
+          <div class="threeD-trade-actions-agreement"> 
+            <input style="cursor: pointer;" type="checkbox" id="threeD-trade-agreement" name="threeD-trade-agreement" ${agreement ? 'checked' : ''} onclick="threeD-trade-agreement">
```
Inline styles should be in the `<style>` instead.

avaer commented on threeD-components/Trade.js:
```diff
@@ -0,0 +1,303 @@
+            <label style="font-size: 20px; cursor: pointer;" for="threeD-trade-agreement" id="threeD-trade-agreement-label"> I agree to trade my token.</label><br>
+            <div class="threeD-spacer"></div>
+            <a class="threeD-trade-cancel" id="threeD-trade-cancel" onclick="threeD-trade-cancel">Cancel</a>
+            <a class="threeD-trade-accept ${!agreement || !toPeer || !fromPeer || !selectedItem ? 'disabled' : ''}" id="threeD-trade-accept" onclick="threeD-trade-accept">Accept</a>
```
This could be abstracted into a local function the names what logic we are trying to capture in this condition.

ChrisEddy replied:
https://github.com/webaverse/app/pull/160/commits/be30d4518b2826fc5f7c3d6fc00120ba9cca4c9a

avaer commented:
The borders on the trade page are poppy when you click to highlight.

avaer commented:
I think the checkmark should clear when you make changes in the trade menu. Otherwise the confirmation isn't really a confirmation of anything, it's just an extra step.

ChrisEddy commented:
> I think the checkmark should clear when you make changes in the trade menu. Otherwise the confirmation isn't really a confirmation of what you selected, it's just an extra step.

fixed.


> The borders on the trade page are poppy when you click to highlight.

fixed.

avaer commented on components/App.js:
```diff
@@ -42,6 +53,98 @@ export const onclickBindings = {
+  },
+  'threeD-trade-header-backBtn': (e) => {
+    const { menu } = getState();
+    menu.trade = JSON.parse(JSON.stringify(defaultState.menu.trade));
```
This could be abstracted in `clone` util method.

ChrisEddy replied:
e90b9bb

avaer commented on components/App.js:
```diff
@@ -42,6 +53,98 @@ export const onclickBindings = {
+  },
+  'threeD-trade-peers-forward': (e) => {
+    const { menu } = getState();
+    menu.trade.peersPage < Math.round(menu.social.peers.length / peersPageLimit) ? menu.trade.peersPage++ : null;
```
Should this be ceil?

ChrisEddy replied:
no we want to round up

avaer commented on components/App.js:
```diff
@@ -42,6 +53,98 @@ export const onclickBindings = {
     });
   },
+  'threeD-social-tradeBtn': (e) => {
+    const { menu } = getState();
```
These braces do not follow lint rules.

ChrisEddy replied:
if your linter is different then normal JS rules then need to lint with script

avaer commented on components/App.js:
```diff
@@ -4,15 +4,26 @@ import uiManager from '../ui-manager.js';
 
+export const isTradeReady = () => {
+  const { menu } = getState();
+  if(menu.trade.agreement && menu.trade.toPeer && menu.trade.fromPeer && menu.trade.selectedItem) {
```
Why not just `return` this?

ChrisEddy replied:
lol sure

avaer commented on threeD-components/Social.js:
```diff
@@ -56,8 +56,8 @@ const Social = (props = {}) => {
-          return `<a class=peer id=peer name=${rig.peerConnection.connectionId}>
-            <div class=name>${rig.textMesh.text}</div>
+          return `<a class=peer id=peer name=${rig.peerConnection?.connectionId}>
+            <div class=name>${rig.textMesh?.text}</div>
```
Can there ever be no `textMesh`? Similar for `peerConnection`.

ChrisEddy replied:
maybe, we don't know

avaer commented on constants.js:
```diff
@@ -55,4 +55,7 @@ export const worldsHost = 'https://worlds.exokit.org';
\ No newline at end of file
+export const localstorageHost = 'https://localstorage.webaverse.com';
+
+export const inventoryPageLimit = 6;
```
I actually don't think these constants are very global. They are very tied to the design of the inventory, so should probably live in the inventory component  files.

avaer reviewed (changes requested):
Mostly stylistic things.

avaer commented on components/App.js:
```diff
@@ -4,67 +4,166 @@ import uiManager from '../ui-manager.js';
 
+export const isTradeReady = () => {
+  const {menu} = getState();
+  return menu.trade.agreement && menu.trade.toPeer && menu.trade.fromPeer && menu.trade.selectedItem;
```
Probably should convert to an actual boolean.
